src/emulation_core/riscv/datapath.rs

```diff
@@ -686,12 +693,21 @@ impl RiscDatapath {
         // be the first register, but the second may be either the
         // second register, the sign-extended immediate value, or the
         // zero-extended immediate value.
-        self.state.alu_input1 = self.state.read_data_1;
+        self.state.alu_input1 = match self.signals.op1_select {
+            OP1Select::PC => self.registers.pc,
+            OP1Select::DATA1 => self.state.read_data_1,
+        };
+
         self.state.alu_input2 = match self.signals.op2_select {
             OP2Select::DATA2 => self.state.read_data_2,
-            OP2Select::IMM => self.state.imm as u64,
+            OP2Select::IMM => self.state.imm as i64 as u64,
```

Comment:
Gonna have to fix some of them since I misinterpreted the Rust docs slightly, but they have to be double case in some cases because I need to make the types a signed integer first, then cast to 64 bits to sign-extend it. But I did make a mistake with making a lot of "as i64 as u64" statements when it should be "as i32 as u64", since the source has to be signed to be sign-extended, not the target. I assumed casting to i64 does the extension and then I just have to respecify as u64, but according to the docs, I should respecify the source as signed and then cast to a u64.

For a lot of these I have to get a u32 to a u64, but via sign extension, so I have to go u32 -> i32 (No-op) -> u64 (Sign extension from a signed type).